### PR TITLE
get tasl from prismic

### PR DESCRIPTION
**Pull request by jamesgorrie**
References #1404 

# Type
✨ Feature  
🐛 Bugfix  

## Value
We can start attributing images correctly.
https://wiki.creativecommons.org/wiki/Best_practices_for_attribution#Title.2C_Author.2C_Source.2C_License

This is a hack, but it's one until Prismic decide on what they're going to do with this.

## Example front-end with different data
__with all information__
screen shot 2017-08-30 at 17 23 58

__without author__
screen shot 2017-08-30 at 17 24 49

__without copyright or source link__
screen shot 2017-08-30 at 17 25 42

### CODE REVIEW

**Comment by jennpb** [comment]
@jamesgorrie I've just been working on this for works, so we can bring some consistency to this if appropriate.

[[Title | Work link]]. Credit: Wellcome Collection [[Licence code | Link]]

example:
'[Astronomy: a diagram of the sun, and various effects of sunlight. Coloured engraving.](https://next.wellcomecollection.org/works/ezyb33ve)' Credit: Wellcome Collection. [CC BY](https://creativecommons.org/licenses/by/4.0/).

**Comment by jamesgorrie** [comment]
@jennpb perfect! This was more the input and was lacking exactly that.
Assuming work link is the S in TASL?

I can think of a case that might come up quite often (not with works but other photography) where we have no title but a source.

**Comment by jennpb** [comment]
@jamesgorrie Yes - source is the URL which we can link from title and image (if image is coming from elsewhere, e.g. works page).

File: server/services/prismic-content.js [hunk 1]
```diff
@@ -44,17 +44,33 @@ export function getFeaturedMediaFromBody(doc): ?Picture {
...
+  // We expect a string of title|author|source|license
+  // e.g. |Rob Bidder||CC-BY-NC
+  const list = copyright.split('|');
+  const v = list
+    .concat(Array(4 - list.length))
+    .map(v => !v.trim() ? null : v);
+
+  const [title, author, source, license] = v;
+
+  return {title, author, source, license};
```
Comment (jamesgorrie) [inline server/services/prismic-content.js]: Could we chuck copyrightHolder and copyrightHolderLink in here too and drop it above?

So we might have something like:
`Ocean views | Greg M | https://wc.org/things/2837 | CC-BY-NC | Wellcome Images | https://wi.org`

File: server/filters/get-tasl-markup.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+import getLicenseInfo from './get-license-info';
+
+export default function({title, author, source, license}) {
+  const titleMarkup = source ? `<a rel="cc:attributionURL" property="dc:title" href="${source}">${title || source}</a>. ` : `<span property="dc:title">${title}.</span> `;
+  const authorMarkup = author ? `Credit: <span property="cc:attributionName">${author}.</span> ` : '';
+  const licenseInfo = license && getLicenseInfo(license);
+  const licenseMarkup = licenseInfo ? `<a rel="license" href="${licenseInfo.url}">${licenseInfo.text}</a>.` : '';
```
Comment (jamesgorrie) [inline server/filters/get-tasl-markup.js]: Could we do this in a nunjucks template or JSX?

File: server/model/picture.js [hunk 3]
```diff
@@ -12,6 +13,7 @@ export type Picture = {|
   copyrightHolder?: string;
   fileFormat?: string;
   url?: ?string;
+  attribute?: {title: ?string, author: ?string, source: ?string, license: LicenseType};
```
Comment (jamesgorrie) [inline server/model/picture.js]: As mentioned - this might be nicer on the root of the object.

File: server/filters/get-tasl-markup.js (hunk 2 above)
Comment (jamesgorrie) [reply server/filters/get-tasl-markup.js]: We don't use this anywhere else, and it would be nice to avoid.

Comment (davidpmccormick) [reply server/filters/get-tasl-markup.js]: I feel it's cleaner here than clogging up the markup with conditionals.

File: server/model/picture.js (hunk 3 above)
Comment (davidpmccormick) [reply server/model/picture.js]: I was inclined to leave it as is – it made sense to me to be grouped together.

File: server/services/prismic-content.js (hunk 1 above)
Comment (davidpmccormick) [reply server/services/prismic-content.js]: Yep, good idea.

File: server/model/picture.js (hunk 3 above)
Comment (jamesgorrie) [reply server/model/picture.js]: I don't think it does - attribution isn't really a type - but rather a way of displaying properties of an image.

File: server/filters/get-tasl-markup.js (hunk 2 above)
Comment (jamesgorrie) [reply server/filters/get-tasl-markup.js]: > clogging up the markup with conditionals

If I search for `{% if` I get 152 results in the markup.
Isn't conditionals what templating is for?

File: server/model/picture.js (hunk 3 above)
Comment (davidpmccormick) [reply server/model/picture.js]: Fair – I'll update.

File: server/filters/get-tasl-markup.js (hunk 2 above)
Comment (davidpmccormick) [reply server/filters/get-tasl-markup.js]: within reason!

File: server/filters/get-tasl-markup.js [hunk 4]
```diff
@@ -1,10 +1,43 @@
...
+  } else {
+    return `Source: ${sourceName}. `;
+  }
+}
+
+function setCopyright(copyrightHolder, copyrightLink) {
+  if (!copyrightHolder) return '';
+
+  if (copyrightLink) {
+    return `Copyright: <a href="${copyrightLink}">${copyrightHolder}</a>. `;
```
Comment (jamesgorrie) [inline server/filters/get-tasl-markup.js]: Could we use the `©` sign?

File: server/filters/get-tasl-markup.js [hunk 5]
```diff
@@ -1,10 +1,43 @@
...
+  if (!sourceName) return '';
+
+  if (sourceLink) {
+    return `Source: <a href="${sourceLink}" rel="cc:attributionURL">${sourceName}</a>. `;
+  } else {
+    return `Source: ${sourceName}. `;
+  }
+}
+
+function setCopyright(copyrightHolder, copyrightLink) {
```
Comment (jamesgorrie) [inline server/filters/get-tasl-markup.js]: Using `set` as the verb here is weird - would something more like `getSourceHtml` make more sense?

File: server/filters/get-tasl-markup.js (hunk 2 above)
Comment (jamesgorrie) [reply server/filters/get-tasl-markup.js]: Again, I really don't like using the another way of doing templating that is quite old school, but as it's my preference, let's get it in and I can use this as an example for using JSX.

File: server/model/picture.js [hunk 6]
```diff
@@ -10,10 +10,9 @@ export type Picture = {|
   alt?: ?string;
   copyright?: ?string;
   author?: string;
-  copyrightHolder?: string;
   fileFormat?: string;
   url?: ?string;
-  attribute?: {title: ?string, author: ?string, source: ?string, license: LicenseType};
+  attribute?: {title: ?string, author: ?string, sourceName: ?string, sourceLink: ?string, license: LicenseType, copyrightHolder: ?string, copyrightLink: ?string};
```
Comment (jamesgorrie) [inline server/model/picture.js]: Can you explain why you wouldn't want this on the root - I can think of many instances where I might just want one bit of this information - it isn't a standardised nor logical grouping, bar for display purposes.

File: server/filters/get-tasl-markup.js (hunk 5 above)
Comment (davidpmccormick) [reply server/filters/get-tasl-markup.js]: I knew something was making me feel weird about this – thanks, have updated.

File: server/filters/get-tasl-markup.js (hunk 4 above)
Comment (davidpmccormick) [reply server/filters/get-tasl-markup.js]: Yo!

File: server/filters/get-tasl-markup.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  if (!title) return '';
+
+  const byAuthor = author ? `, by ${author}` : '';
+
+  if (sourceLink) {
+    return `<a href="${sourceLink}" property="dc:title" rel="cc:attributionURL">${title}${byAuthor}</a>. `;
+  } else {
+    return `<span property="dc:title">${title}${byAuthor}.</span> `;
+  }
+}
```
Comment (jamesgorrie) [inline server/filters/get-tasl-markup.js]: Way clearer 👓

File: server/model/picture.js [hunk 8]
```diff
@@ -9,9 +10,13 @@ export type Picture = {|
...
   copyright?: ?string;
   author?: string;
-  copyrightHolder?: string;
   fileFormat?: string;
   url?: ?string;
+  title?: ?string;
+  author?: ?string;
+  source?: ?{name?: ?string, link?: ?string};
+  license?: ?LicenseType;
+  copyright?: ?{holder?: string, link?: string};
```
Comment (jamesgorrie) [inline server/model/picture.js]: It'd be good to stop using double nullables...

I think having `license: ?LicenceType` is better, forcing someone to purposefully pass `null` if they don't have it. Just a thought for discussion.

File: server/services/prismic-content.js [hunk 9]
```diff
@@ -44,17 +44,42 @@ export function getFeaturedMediaFromBody(doc): ?Picture {
...
+
   return ({
     type: 'picture',
     contentUrl: image && convertPrismicToImgIxUri(image.url), // TODO: Send this through the img.wc.org
     width: image && image.dimensions.width,
     height: image && image.dimensions.height,
     caption: captionedImage.caption && captionedImage.caption.length !== 0 && asHtml(captionedImage.caption),
     alt: image && image.alt,
-    copyrightHolder: image && image.copyright
+    title: tasl && tasl.title,
```
Comment (jamesgorrie) [inline server/services/prismic-content.js]: Seriously considering the Maybe monad:
https://monet.github.io/monet.js/#maybe

File: server/services/prismic-content.js [hunk 10]
```diff
@@ -44,17 +44,42 @@ export function getFeaturedMediaFromBody(doc): ?Picture {
...
+    author: tasl && tasl.author,
+    source: {
+      name: tasl && tasl.sourceName,
+      link: tasl && tasl.sourceLink
+    },
+    license: tasl && tasl.license,
+    copyright: {
+      holder: tasl && tasl.copyrightHolder,
+      link: tasl && tasl.copyrightLink
```
Comment (jamesgorrie) [inline server/services/prismic-content.js]: I don't really mind, but why not have `copyrightHolder` and `copyrightLink`?

**Comment by jamesgorrie** [comment]
Are we still in `WIP` because it looks good to me?

**Comment by jamesgorrie** [comment]
@davidpmccormick have we looked to see what happens with existing credits?

**Comment by jamesgorrie** [comment]
> existing credits

Maybe if there is just 1 in the array - we just spit that out?

**Comment by davidpmccormick** [comment]
> have we looked to see what happens with existing credits?

It will just spit it out, currently. So where we currently have 'Credit: Wellcome Collection', we will have 'Wellcome Collection'. This is probably OK?

**Comment by jamesgorrie** [comment]
> This is probably OK?

More than.

**Pull request merged**